File: src/learn/ds/array/search/BinarySearch.java
package learn.ds.array.search;

/**
 * @author Varma Penmetsa
 *
 * https://www.geeksforgeeks.org/binary-search/
 */
public class BinarySearch {

    /**
     * Recursive implementation
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public static int search(int array[],int x, int left, int right) {
        if (right >= left) {
            int mid = left + (right - left) / 2;

            if (array[mid] == x) {
                return mid;
            }else if (x < array[mid]) {
                return search(array, x, left, mid - 1);
            }else {
                return search(array, x, mid + 1, right);
            }
        }
        return -1;
    }

    /**
     * Iterative Approach
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public static int search(int[] array, int x){
        int left = 0;
        int right = array.length - 1;
        int mid;

        while (left <= right){
            mid = left + (right - left) / 2;
            if(array[mid] == x){
                return mid;
            }else if(array[mid] < x){
                left = mid + 1;
            }else {
                right = mid -1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
        int n = arr.length;
        System.out.println(search(arr, 9));
    }
}

File: python/algo/search/BinarySearch.py
from __future__ import print_function

class BinarSearch(object):

    #Iterative Binary Search on Arrays
    def search1(self, array, x):
        left = 0
        right = len(array) - 1

        while(left <= right):
            mid  = left + (right - left) / 2
            if(array[mid] == x):
                return mid
            elif(array[mid] < x):
                left = mid + 1
            else:
                right = mid - 1
        return -1

    # Binary Search Recursive
    def search(self, array, x, left, right):
        if(left > right):
            return -1

        mid = left + (right - left) / 2

        if(array[mid] == x):
            return mid
        elif(array[mid] < x):
            return self.search(array, x, mid + 1, right)
        else:
            return self.search(array, x, left, mid -1)

if __name__ == '__main__':

    b = BinarSearch()
    array = [1, 2, 4, 5, 6, 7, 9]

    #Iterative
    print(b.search1(array,0))

    #Recursive
    print(b.search(array, 0, 0, len(array)-1 ))
